### DISCUSSION TITLE
Native Image support for Java Agent

### DISCUSSION THREAD

**Discussion by yswdqz** [oracle/graal#6693] [Q&A] [answered]
Hi 👋
I am currently undertaking research on native-image support for Java Agents. However, I have encountered numerous challenges and feel uncertain about how to proceed. 

I've read #1065, and other issues about javaagent. It appears that javaagent is not officially supported. Does this mean that I can go about getting native-image to support javaagent in my own way?

Thank you for considering my request. Any help is really appreciated !

**Comment by christianwimmer** [comment]
We are working on basic support for JVMTI agents (so agents written in C). We can support some of the JVMTI capabilities there, but certainly not any kind of bytecode transformation.

But I assume with "Java Agents" you mean the [Java instrumentation API](https://docs.oracle.com/en/java/javase/17/docs/api/java.instrument/java/lang/instrument/package-summary.html). This API does not really support much more than bytecode transformation, and since that cannot be done at image run time, there is not much point in loading such agents at run time.

All bytecode transformation must be done at image build time.

**Reply by yswdqz** [reply]
Thank you for your help !
 I also found that `Java on Truffle` can run in the context of a native executable while still allowing dynamically-loaded bytecodes. Can I use it to support bytecode transformation at run time ?

**Reply by christianwimmer** [reply]
Truffle is most likely not what you want, because then your application is no longer AOT compiled.

**Reply by yswdqz** [reply]
I understand. For my case, it might be better to write a class to implement `org.graalvm.nativeimage.hosted.Feature` to add the transformed classes to the predefined-class list?

**Comment by vjovanov** [comment, marked as answer]
The bytecode transformation is best done by passing a Java agent at build time with `-J-javaagent:agent.jar` to `native-image`. There are a few things to note: 
 - The agent is running and transforming all classes at build time.
 - The agent should not touch the classes that are used by `native-image` itself. One can get in trouble then as unpredictable things can happen.
 - The `premain` of the agent will be executed at build time.
 - The agent is currently marked as `link-at-build-time` so all of the classes need to be present in the agent jar.

If you encounter issues please contact me via our [community Slack](https://graalvm.slack.com/join/shared_invite/zt-wludoxhy-ZSoK2rYEt5QC_FzSf5N~~A#/shared-invite/email).
